### PR TITLE
feat: allow api key and keychain notarize

### PR DESCRIPTION
**Pull request by mifi**
see also #7859

I've tested this here https://github.com/mifi/lossless-cut/actions/runs/6719326573/job/18260738394
 with a [patched macPackager.js](https://github.com/mifi/lossless-cut/blob/master/.yarn/patches/app-builder-lib-npm-24.8.0-51e1f5cd3f.patch)

### CODE REVIEW

**Comment by mmaietta** [comment]
`tsc` build step is failing with these errors:
```
Error: packages/app-builder-lib/src/macPackager.ts(484,36): error TS7006: Parameter 'appPath' implicitly has an 'any' type.
Error: packages/app-builder-lib/src/macPackager.ts(504,54): error TS2322: Type 'string | undefined' is not assignable to type 'string'.
  Type 'undefined' is not assignable to type 'string'.
Error: packages/app-builder-lib/src/macPackager.ts(504,67): error TS2322: Type 'string | undefined' is not assignable to type 'string'.
  Type 'undefined' is not assignable to type 'string'.
Error: packages/app-builder-lib/src/macPackager.ts(504,82): error TS2322: Type 'string | undefined' is not assignable to type 'string'.
  Type 'undefined' is not assignable to type 'string'.
```
Can you please add explicit types to the consts?

File: packages/app-builder-lib/src/macPackager.ts [hunk 1]
```diff
@@ -481,31 +481,55 @@ export default class MacPackager extends PlatformPackager<MacConfiguration> {
...
+      }
+      return this.generateNotarizeOptions({ appPath, appleId, appleIdPassword })
+    }
+
+    // option 2: API key
+    const appleApiKey = process.env.APPLE_API_KEY;
+    const appleApiKeyId = process.env.APPLE_API_KEY_ID;
+    const appleApiIssuer = process.env.APPLE_API_ISSUER;
+    if (appleApiKey || appleApiKeyId || appleApiIssuer) {
+      return this.generateNotarizeOptions({ appPath, appleApiKey, appleApiKeyId, appleApiIssuer });
```
Comment (mmaietta) [inline packages/app-builder-lib/src/macPackager.ts:504]: Do these need to be `&&` in order for it to be of type `string`?

File: packages/app-builder-lib/src/macPackager.ts [hunk 2]
```diff
@@ -481,31 +481,55 @@ export default class MacPackager extends PlatformPackager<MacConfiguration> {
...
-    }
-    if (!appleId) {
-      throw new InvalidConfigurationError(`APPLE_ID env var needs to be set`)
-    }
-    if (!appleIdPassword) {
-      throw new InvalidConfigurationError(`APPLE_APP_SPECIFIC_PASSWORD env var needs to be set`)
-    }
-    const options = this.generateNotarizeOptions(appPath, appleId, appleIdPassword)
+    const options = this.getNotarizeOptions(appPath)
+    if (options == null) return
```
Comment (mmaietta) [inline packages/app-builder-lib/src/macPackager.ts:525]: Please make this multi-line to keep coding styles consistent
```
if (options == null) {
   return
}
```

**Comment by mifi** [comment]
fixed some types. however i'm really struggling to fix the types for `generateNotarizeOptions`. when `notarize: true` i'm struggling to distinguish between legacy and non legacy options. and if I don't distinguish, it complains that `notarytool` option is missing, which makes sense because @electron/notarize types [requires `tool` to be `'notarytool'`](https://github.com/electron/notarize/blob/c9af67d0dcbc1d4ba86cf6108c810e828d1adefd/src/types.ts#L62), which contradicts with their readme which says:

> tool String - The notarization tool to use, default is notarytool

not sure if i'm misunderstanding something

**Comment by bcomnes** [comment]
This is a great improvement, however at the moment, it isn't functional. I opened a small fix to make it work: https://github.com/electron-userland/electron-builder/pull/7951

tldr The notarization tool never runs when the API key is set and notarize is set to `true`, and there are no other workarounds that work without the fix. Hope we can land it!
